source/core/src/main/com/csse3200/game/components/inventory/PlayerInventoryDisplay.java

```diff
@@ -82,28 +85,78 @@ public PlayerInventoryDisplay(Inventory inventory, int numCols, int hotBarCapaci
+        } else { // Toggle on
             logger.debug("Inventory toggled on.");
             generateInventory();
-            stage.getActors().removeValue(hotBarDisplay, true); // close hot-bar
-            InventoryUtils.disposeGroupRecursively(hotBarDisplay);
-            toggle = true;
         }
+        // Regenerate the hotBar in either case
+        stage.getActors().removeValue(hotBarDisplay, true);
+        InventoryUtils.disposeGroupRecursively(hotBarDisplay);
+        generateHotBar();
+    }
+
+    /**
+     * Sets up drag-and-drop functionality for the inventory system using the provided slot and item.
+     * The method defines both the drag source and the drop target, allowing users to drag items between slots.
+     * The items are swapped between the source and target slots upon successful drop.
+     *
+     * @param slot        The {@link ImageButton} that represents the inventory slot where the item is being dragged from or dropped into.
+     * @param targetIndex The target index of the slot in the inventory where the item will be dropped.
+     * @param item        The {@link AbstractItem} representing the item in the source slot being dragged.
+     */
+    private void setupDragAndDrop (ImageButton slot,int targetIndex, AbstractItem item) {
+        // Define the source
+        dragAndDrop.addSource(new DragAndDrop.Source(slot) {
+            @Override
+            public DragAndDrop.Payload dragStart(InputEvent event, float x, float y, int pointer) {
+                DragAndDrop.Payload payload = new DragAndDrop.Payload();
+                if(item != null) {
+                    payload.setObject(targetIndex);
+                    Image draggedImage = new Image(new Texture(item.getTexturePath()));
+                    draggedImage.setSize(80, 80);
+                    payload.setDragActor(draggedImage);
+                }
+                return payload;
+            }
+        });
+
+        // Define the target
+        dragAndDrop.addTarget(new DragAndDrop.Target(slot) {
+            @Override
+            public boolean drag(DragAndDrop.Source source, DragAndDrop.Payload payload, float x, float y, int pointer) {
+                // Optional: Highlight the target slot to indicate a valid drop zone
+                getActor().setColor(Color.LIGHT_GRAY);
+                return true; // Return true to indicate the slot is a valid target
+            }
+
+            @Override
+            public void reset(DragAndDrop.Source source, DragAndDrop.Payload payload) {
+                // Reset the color of the slot when dragging is reset
+                getActor().setColor(Color.WHITE);
+            }
+
+            @Override
+            public void drop(DragAndDrop.Source source, DragAndDrop.Payload payload, float x, float y, int pointer) {
+                int sourceIndex = (int) payload.getObject();
+                    inventory.swap(sourceIndex,targetIndex);
+                    regenerateDisplay();
```

Comment:
Is it necessary to redraw the inventory every time you swap an item? Seems excessive, correct me tho if this is the only way

Reply:
It's not, but this game is so small in terms of memory and speed that anything that's not egregiously slow is irrelevant (the only thing that really takes time in this game is loading assets). There's no point hyper-optimising when this works extremely fast for all practical purposes.

Points 1 & 3 above are irrelevant to this PR - please put them in the team 3 discord chat as design choices that should be made. Regarding point 3 - it's probably more important to make the dialogue box be drawn at a higher Z_index first rather than just moving the inventory (which should be done anyway and the hotbar & inventory designs will be tweaked based on user testing in sprint 4).

As to point 2, we can stick the capacity in the subscript as well if that is what the design committee wants. I believe however, the idea is in future to have more information regarding item's use cases be in the dialogue box or be presented elsewhere.